File: packages/shared/src/validation/environments-validation.ts
import type { EnvironmentsConfig, Environment } from "@shared/types";

export type EditableEnv = {
  id: string;
  name: string;
  prometheus: string;
  loki: string;
  kubernetes: string;
  aws: string;
};

const NAME_RE = /^[A-Za-z0-9._-]+$/;
const RESERVED = new Set(["test"]);

const REF_FIELDS: { field: keyof EditableEnv; type: string }[] = [
  { field: "prometheus", type: "prometheus" },
  { field: "loki", type: "loki" },
  { field: "kubernetes", type: "kubernetes" },
  { field: "aws", type: "aws" },
];

/** Build the API payload from editable state. */
export function toEnvConfig(envs: EditableEnv[]): EnvironmentsConfig {
  return {
    environments: envs.map((e) => {
      const out: Environment = { name: e.name.trim() };
      if (e.prometheus) out.prometheus = e.prometheus;
      if (e.loki) out.loki = e.loki;
      if (e.kubernetes) out.kubernetes = e.kubernetes;
      if (e.aws) out.aws = e.aws;
      return out;
    }),
  };
}

/** Convert a loaded API config into editable state. */
export function fromEnvConfig(cfg: EnvironmentsConfig | undefined): EditableEnv[] {
  return (cfg?.environments ?? []).map((e, i) => ({
    id: `env-${i}-${e.name ?? ""}`,
    name: e.name ?? "",
    prometheus: e.prometheus ?? "",
    loki: e.loki ?? "",
    kubernetes: e.kubernetes ?? "",
    aws: e.aws ?? "",
  }));
}

export type EnvValidation = { valid: boolean; envs: Record<string, string> };

/** Validate env names + that each endpoint ref exists and is of the right type. */
export function validateEnvironments(
  envs: EditableEnv[],
  endpointsByType: Record<string, string>
): EnvValidation {
  const result: EnvValidation = { valid: true, envs: {} };
  const seen = new Set<string>();

  for (const e of envs) {
    let err = "";
    const name = e.name.trim();
    if (!name) err = "Name is required.";
    else if (!NAME_RE.test(name)) err = "Name may only contain letters, digits, '.', '_' and '-'.";
    else if (RESERVED.has(name)) err = `"${name}" is reserved.`;
    else if (seen.has(name)) err = `Duplicate name "${name}".`;
    else seen.add(name);

    for (const { field, type } of REF_FIELDS) {
      const ref = String(e[field] ?? "");
      if (!ref) continue;
      const actual = endpointsByType[ref];
      if (!actual) err = err || `${type} endpoint "${ref}" no longer exists.`;
      else if (actual !== type) err = err || `${type} endpoint "${ref}" is a ${actual} endpoint.`;
    }

    if (err) {
      result.valid = false;
      result.envs[e.id] = err;
    }
  }

  return result;
}

File: alert-agent/services/environments_validation.py
"""Validate the environment map submitted via the Web UI.

Environments no longer match on labels — each names a set of endpoint refs
(one per source) and maps to a webhook path (``/webhook/<name>``). Refs are
checked against the current endpoint registry when it is supplied.
"""

import re
from typing import Any

# Environment names appear in the webhook URL path, so keep them path-safe.
_NAME_RE = re.compile(r"^[A-Za-z0-9._-]+$")
# Reserved: /webhook/test is the (auth-gated) test route.
_RESERVED_NAMES = {"test"}

# ref field on an environment → endpoint type it must point at.
_REF_TYPES = {
    "prometheus": "prometheus",
    "loki": "loki",
    "kubernetes": "kubernetes",
    "aws": "aws",
}


def validate_environments_body(
    body: dict[str, Any],
    endpoints_by_type: dict[str, str] | None = None,
) -> list[str]:
    """Return a list of human-readable validation errors (empty if valid).

    ``endpoints_by_type`` maps endpoint name → type; when provided, each ref is
    verified to exist and match the expected type.
    """
    errors: list[str] = []

    environments = body.get("environments")
    if environments is None or not isinstance(environments, list):
        errors.append("environments must be a list.")
        return errors

    seen_names: set[str] = set()
    for i, env in enumerate(environments, start=1):
        if not isinstance(env, dict):
            errors.append(f"Environment {i} must be an object.")
            continue

        name = str(env.get("name") or "").strip()
        prefix = f"Environment {i}" + (f" ({name})" if name else "")
        if not name:
            errors.append(f"{prefix}: name is required.")
        elif not _NAME_RE.match(name):
            errors.append(f"{prefix}: name may only contain letters, digits, '.', '_' and '-'.")
        elif name in _RESERVED_NAMES:
            errors.append(f"{prefix}: {name!r} is reserved.")
        elif name in seen_names:
            errors.append(f"{prefix}: duplicate name {name!r}.")
        else:
            seen_names.add(name)

        for ref_field, expected_type in _REF_TYPES.items():
            ref = env.get(ref_field)
            if ref in (None, ""):
                continue
            ref = str(ref)
            if endpoints_by_type is None:
                continue
            actual = endpoints_by_type.get(ref)
            if actual is None:
                errors.append(f"{prefix}: {ref_field} endpoint {ref!r} does not exist.")
            elif actual != expected_type:
                errors.append(
                    f"{prefix}: {ref_field} endpoint {ref!r} is a {actual} endpoint, expected {expected_type}."
                )

    return errors
